nixos/zerotierone: Update Zerotier's Nix module to include to ability to define local.conf in Nix

yonghan opened the pull request:
Added option to write zerotier's local.conf in nix.

## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

yonghan committed:
Update zerotierone.nix

Added option to write zerotier's local.conf in nix.

yonghan committed:
Update zerotierone.nix

yonghan committed:
Update zerotierone.nix

yonghan committed:
Merge branch 'NixOS:master' into yonghan-zerotier

yonghan committed:
Update zerotierone.nix

yonghan commented:
@ehmry can you also look at https://github.com/NixOS/nixpkgs/pull/187373? It is similar to this PR.

ehmry commented on nixos/modules/services/networking/zerotierone.nix:
```diff
@@ -29,6 +31,19 @@ in
+    example = {
+      settings.allowTcpFallbackRelay = false;
+    };
+    type = types.attrs;
```
Actually I think it would be better to make the default `null`, the type `nullOr attrs`, and don't touch the file in `/var/lib` unless `localConf` is not null. This way nothing changes unless the option is set.

yonghan replied:
Good point. Let me see how I can change that.

yonghan replied:
Done!

yonghan replied:
Done!

yonghan committed:
Merge branch 'NixOS:master' into yonghan-zerotier

yonghan committed:
Update zerotierone.nix per input

Changed `services.zerotierone.localConf` default value to `null` and type to `nullOr attrs` per input.

Changed `systemd` preStart script to delete
 `/var/lib/zerotier-one/local.conf` when it is a symlink, and rename when it is an actual file, then only create a symlink to the nix store when `services.zerotierone.localConf` is not null.

yonghan committed:
delete trailing whitespace at row 70

yonghan committed:
Fix bash prestart script syntax error

ehmry commented on nixos/modules/services/networking/zerotierone.nix:
```diff
@@ -45,7 +60,18 @@ in
+        then
+          mv ${localConfFilePath} ${localConfFilePath}.bak
+        fi
+      '' + optionalString (cfg.localConf != null) ''
```
If `cfg.localConf == null` then I don't think anything in /var/lib/zerotier-one should modified.

yonghan replied:
Fixed

yonghan committed:
Merge branch 'NixOS:master' into yonghan-zerotier

yonghan committed:
Update zerotierone.nix per input

Nothing will happen now when `cfg.localConf == null`

stigtsp commented:
Hi there! Some nitpicks:

Just noticed that this PR has been merged with several commits messages like:

5fff7f40788a Update zerotierone.nix per input
7b83a839dc9f Fix bash prestart script syntax error
996bbe5bd96e delete trailing whitespace at row 70
8af401d3cfc4 Update zerotierone.nix per input
b233faab8289 Update zerotierone.nix
36791bab6a50 Update zerotierone.nix
0b357293e8b1 Update zerotierone.nix
097f2b673741 Update zerotierone.nix

Don't think this fits with [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md)
